### PR TITLE
Convert `as` commands to the rzshell

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- `asj` now became `aslj`
- `ask` was removed
- `as <name>` became `asn <name>`
- `as`, `asca`, `asc` take now only the number as the argument

I have a question though - should I ignore the name argument in `asca` and `asc` completely, or return it back and allow the argument to be **either** syscall number or syscall name?

**Test plan**

CI is green

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/1442

### CODE REVIEW

**Comment by ret2libc** [comment]
> I have a question though - should I ignore the name argument in asca and asc completely, or return it back and allow the argument to be either syscall number or syscall name?

I think it's oki to accept just a string and recognize the type of argument, whether number or string.

File: librz/core/cmd/cmd_analysis.c [hunk 1]
```diff
@@ -3942,131 +3841,118 @@ static void cmd_sdbk(Sdb *db, const char *input) {
...
+	if (!sysc) {
+		RZ_LOG_ERROR("Cannot resolve syscall: %" PFMT64d "\n", num);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	rz_cons_println(sysc);
+	free(sysc);
+	return RZ_CMD_STATUS_OK;
+}
+
+static RzCmdStatus syscall_dump(RzCore *core, int argc, const char **argv, bool is_c) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: please simplify this function, i think it should be quite straightforward

File: librz/core/csyscall.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * Given the syscall number and address it resolves the syscall
+ * for the selected `asm.arch` and `asm.os` values and print
+ * its arguments.
+ *
+ * \param core RzCore instance
+ * \param n number of the syscall
+ * \param addr address of the syscall
+ */
+RZ_API RZ_OWN char *rz_core_syscall_as_string(RzCore *core, st64 n, ut64 addr) {
```
Comment (ret2libc) [inline librz/core/csyscall.c:26]: this function is such a mess :( i understand you just copy-pasted it, so don't worry

File: librz/core/cmd/cmd_analysis.c [hunk 3]
```diff
@@ -3942,131 +3841,118 @@ static void cmd_sdbk(Sdb *db, const char *input) {
...
+	return sdb_fmt(n > 1000 ? "0x%x" : "%d", n);
+}
+
+RZ_IPI RzCmdStatus rz_analysis_syscall_print_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
 	RzListIter *iter;
-	RzList *list;
-	RzNum *num = NULL;
-	int n;
+	RzSyscallItem *si;
+	RzList *list = rz_syscall_list(core->analysis->syscall);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:3016]: move this to a separate function in csyscall.c imo. you avoid having syscallNumber function both here and there. Also, move the other function that uses syscallNumber there.

File: librz/core/csyscall.c (hunk 2 above)
Comment (XVilka) [reply librz/core/csyscall.c:26]: And it seems not working, `rz_core_syscall_as_string(core, 4, 0x0)` returns `read(0)`, not `stat()`.

File: librz/core/cmd_descs/cmd_analysis.yaml [hunk 4]
```diff
@@ -1453,3 +1453,49 @@ commands:
...
+        cname: analysis_syscall_name
+        args:
+          - name: name
+            type: RZ_CMD_ARG_TYPE_STRING
+      - name: asr
+        summary: Returns the syscall name by the number
+        cname: analysis_syscall_number
+        args:
+          - name: number
+            type: RZ_CMD_ARG_TYPE_STRING
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml]: this is string here but i think it is just converted to a number in the handler.

File: librz/core/cmd_descs/cmd_analysis.yaml [hunk 5]
```diff
@@ -1453,3 +1453,49 @@ commands:
...
+            optional: true
+      - name: asl
+        summary: List syscalls
+        cname: analysis_syscall_print
+        type: RZ_CMD_DESC_TYPE_ARGV_STATE
+        default_mode: RZ_OUTPUT_MODE_STANDARD
+        modes:
+          - RZ_OUTPUT_MODE_STANDARD
+          - RZ_OUTPUT_MODE_JSON
+        args: []
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_analysis.yaml:1859]: In the past it was possible to list just one syscall. I think we can keep this feature or actually we could even accept an array of strings and return all syscalls matching the array.

File: librz/core/csyscall.c (hunk 2 above)
Comment (wargio) [reply librz/core/csyscall.c:26]: maybe let's open an issue so we track that this does not work.

File: librz/core/cmd_descs/cmd_analysis.yaml (hunk 5 above)
Comment (XVilka) [reply librz/core/cmd_descs/cmd_analysis.yaml:1859]: I moved this feature into the `asn` and `asr` commands respectively.

File: librz/core/cmd/cmd_analysis.c (hunk 3 above)
Comment (XVilka) [reply librz/core/cmd/cmd_analysis.c:3016]: `syscallNumber()` is used in multiple places and is short enough to be duplicated. Further refactoring is out of the scope of this PR, IMHO. A lot of things has to be changed in how syscall are handled.

File: librz/core/csyscall.c [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * \brief Returns the syscall representation as a string
+ *
+ * Given the syscall number and address it resolves the syscall
+ * for the selected `asm.arch` and `asm.os` values and print
+ * its arguments.
+ *
+ * \param core RzCore instance
+ * \param n number of the syscall
```
Comment (ret2libc) [inline librz/core/csyscall.c]: there is no mention here about what this function does when `n` is `-1` or whether it is even valid (according to the code, it is valid). Please add doc for that so we don't have to read the function to know it's an expected value.

File: librz/core/cmd/cmd_analysis.c [hunk 7]
```diff
@@ -3095,143 +2994,122 @@ RZ_API void rz_cmd_analysis_calls(RzCore *core, const char *input, bool printCom
...
+static RzCmdStatus syscalls_dump(RzCore *core, int argc, const char **argv, bool is_c) {
+	if (argc > 1) {
+		st64 n = rz_num_math(core->num, argv[1]);
+		if (n < 1) {
+			n = rz_syscall_get_num(core->analysis->syscall, argv[1]);
+			if (n == -1) {
+				RZ_LOG_ERROR("Cannot resolve syscall: %s\n", argv[1]);
+				return RZ_CMD_STATUS_ERROR;
 			}
+			rz_cons_printf(".equ SYS_%s %s\n", argv[1], syscallNumber(n));
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c:3083]: use `syscall_dump`

File: librz/core/cmd/cmd_analysis.c [hunk 8]
```diff
@@ -3095,143 +2994,122 @@ RZ_API void rz_cmd_analysis_calls(RzCore *core, const char *input, bool printCom
...
+	if (num < 1) {
+		RZ_LOG_ERROR("Cannot resolve syscall: %s\n", argv[1]);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	rz_cons_println(syscallNumber(num));
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_analysis_syscall_number_handler(RzCore *core, int argc, const char **argv) {
+	st64 num = rz_num_math(core->num, argv[1]);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_analysis.c]: AFAIR we should use `rz_num_math(NULL, ...)` when we don't want resolution of flags, function names, etc. like in this case (it is a _NUM argument, not a _RZNUM).

File: librz/core/cmd/cmd_analysis.c (hunk 7 above)
Comment (XVilka) [reply librz/core/cmd/cmd_analysis.c:3083]: I can't since it doesn't search the `RzSyscallItem *si` here, just the `char *` string.

Comment (ret2libc) [reply librz/core/cmd/cmd_analysis.c:3083]: actually... can't we just remove these two lines of code and let the code fall through the `rz_syscall_get`, etc. ?

Comment (XVilka) [reply librz/core/cmd/cmd_analysis.c:3083]: @ret2libc at practice the logic between `rz_syscall_get_num()` and `rz_syscall_get()` is a bit different. I would prefer not to touch too much of this code until we decide to refactor the `librz/syscall/*` code which is a mess.

File: librz/core/csyscall.c (hunk 2 above)
Comment (XVilka) [reply librz/core/csyscall.c:26]: Will create an issue once this merged.

**Review by ret2libc — APPROVED** [review]

File: librz/core/csyscall.c (hunk 2 above)
Comment (XVilka) [reply librz/core/csyscall.c:26]: https://github.com/rizinorg/rizin/issues/2583

**Pull request merged**